### PR TITLE
Inline asm hir pipeline

### PR DESCRIPTION
**Pull request by badumbatish**
For https://github.com/Rust-GCC/gccrs/issues/3102

### CODE REVIEW

**Review by P-E-P — COMMENTED** [review]
Few nits but it looks mostly good :smiley:

File: gcc/rust/expand/rust-macro-builtins-asm.cc [hunk 1]
```diff
@@ -378,17 +391,21 @@ parse_reg_operand_inout (InlineAsmContext inline_asm_ctx)
       // TODO: Is error propogation our top priority, the ? in rust's asm.rs is
       // doing a lot of work.
       // TODO: Not sure how to use parse_expr
-      auto expr = parse_format_string (inline_asm_ctx);
+      auto exist_ident1 = check_identifier (parser, "");
+      if (!exist_ident1)
```
[the remark this answers is not in the document]
Comment (badumbatish) [reply gcc/rust/expand/rust-macro-builtins-asm.cc:395]: will do

File: gcc/rust/expand/rust-macro-builtins-asm.cc [hunk 2]
```diff
@@ -227,20 +227,28 @@ parse_reg_operand (InlineAsmContext inline_asm_ctx)
...
+	  //
+	  inline_asm_ctx = *result;
+	  break;
+	}
+      else if (result.error () == COMMITTED
+	       && parse_func != parse_reg_operand_unexpected)
+	return result;
+      else if (result.error () == COMMITTED
+	       && parse_func == parse_reg_operand_unexpected)
+	return inline_asm_ctx;
```
[the remark this answers is not in the document]
Comment (badumbatish) [reply gcc/rust/expand/rust-macro-builtins-asm.cc:251]: will resolve this with your first approach, ty!

File: gcc/rust/hir/tree/rust-hir-expr.h [hunk 3]
```diff
@@ -3843,28 +3843,278 @@ class InlineAsmRegClass
...
+  tl::optional<struct Label> label;
+
+  InlineAsmOperand () {}
+  InlineAsmOperand (const InlineAsmOperand &other)
+    : register_type (other.register_type), in (other.in), out (other.out),
+      in_out (other.in_out), split_in_out (other.split_in_out),
+      cnst (other.cnst), sym (other.sym)
+  {}
+
+  void set_in (const struct In &reg)
```
[the remark this answers is not in the document]
Comment (badumbatish) [reply gcc/rust/hir/tree/rust-hir-expr.h:4072]: gotcha, will refactor

File: gcc/rust/hir/rust-ast-lower-expr.cc [hunk 4]
```diff
@@ -829,6 +829,99 @@ ASTLoweringExpr::visit (AST::ClosureExprInnerTyped &expr)
...
+
+  /*In,*/
+  /*Out,*/
+  /*InOut,*/
+  /*SplitInOut,*/
+  /*Const,*/
+  /*Sym,*/
+  /*Label,*/
+  struct HIR::InlineAsmOperand op;
+  if (type == AST::InlineAsmOperand::RegisterType::In)
```
[the remark this answers is not in the document]
Comment (badumbatish) [reply gcc/rust/hir/rust-ast-lower-expr.cc]: Done

File: gcc/rust/hir/rust-ast-lower-expr.cc [hunk 5]
```diff
@@ -829,6 +829,92 @@ ASTLoweringExpr::visit (AST::ClosureExprInnerTyped &expr)
...
+				   std::unique_ptr<Expr> (
+				     ASTLoweringExpr::translate (
+				       *label_value.expr.get ())));
+      return label;
+    }
+  else
+    {
+      rust_unreachable ();
+    }
+}
```
Comment (CohenArthur) [inline gcc/rust/hir/rust-ast-lower-expr.cc:917]: this is great, this is exactly what we want! I would suggest creating a separate function for each type and returning the result of calling it directly instead, which should result in a little less code. something like this 

```c++
switch (get_register_type()) {
    case In:
        return translate_in(...);
    case InOut:
        return translate_in_out(...);
    ...
}
```

File: gcc/rust/hir/rust-ast-lower-expr.cc [hunk 6]
```diff
@@ -837,10 +923,22 @@ ASTLoweringExpr::visit (AST::InlineAsm &expr)
 				 mappings.get_next_hir_id (crate_num),
 				 mappings.get_next_localdef_id (crate_num));
 
+  std::vector<HIR::InlineAsmOperand> hir_operands;
+  std::vector<AST::InlineAsmOperand> ast_operands = expr.get_operands ();
```
Comment (CohenArthur) [inline gcc/rust/hir/rust-ast-lower-expr.cc:927]:
```suggestion
  const std::vector<AST::InlineAsmOperand> &ast_operands = expr.get_operands ();
```

we should be able to just iterate on references

Comment (badumbatish) [reply gcc/rust/hir/rust-ast-lower-expr.cc]: resolve :)

File: gcc/rust/hir/rust-ast-lower-expr.cc (hunk 5 above)
Comment (badumbatish) [reply gcc/rust/hir/rust-ast-lower-expr.cc]: ah i forgot the switch case also, fixed

**Comment by badumbatish** [comment]
fixing this rn

**Review by P-E-P — APPROVED** [review]

**Pull request merged**